pallets/delegation/src/delegation_hierarchy.rs

```diff
@@ -52,87 +57,158 @@ impl Default for Permissions {
+		/// * parent - the parent node ID this node will be a child of
+		/// * owner: the identifier of the owner of the new delegation, i.e.,
+		///   the new delegate
+		/// * permissions: the permission flags for the operations the delegate
+		///   is allowed to perform
+		pub fn new_node_child(
+			root_id: DelegationNodeIdOf<T>,
+			parent: DelegationNodeIdOf<T>,
+			owner: DelegatorIdOf<T>,
+			permissions: Permissions,
+		) -> Self {
+			DelegationNode {
+				root_id,
+				parent: Some(parent),
+				owner,
+				permissions,
+				revoked: false,
+			}
+		}
+	}
 }
 
-impl<T: Config> DelegationNode<T> {
-	/// Create a new delegation node that is a direct descendent of the
-	/// given root.
-	///
-	/// * root_id: the root node ID this node will be a child of
-	/// * owner: the identifier of the owner of the new delegation, i.e., the
-	///   new delegate
-	/// * permissions: the permission flags for the operations the delegate is
-	///   allowed to perform
-	pub fn new_root_child(root_id: DelegationNodeIdOf<T>, owner: DelegatorIdOf<T>, permissions: Permissions) -> Self {
-		DelegationNode {
-			root_id,
-			owner,
-			permissions,
-			revoked: false,
-			parent: None,
+// The new types.
+pub(crate) mod v1 {
+	use super::*;
+
+	#[derive(Clone, Debug, Encode, Decode, PartialEq)]
+	pub struct DelegationNode<T: Config> {
+		pub hierarchy_root_id: DelegationNodeIdOf<T>,
+		pub parent: Option<DelegationNodeIdOf<T>>,
+		pub children: BTreeSet<DelegationNodeIdOf<T>>,
+		pub details: DelegationDetails<T>,
+	}
+
+	impl<T: Config> DelegationNode<T> {
+		pub fn new_root_node(id: DelegationNodeIdOf<T>, details: DelegationDetails<T>) -> Self {
+			Self {
+				hierarchy_root_id: id,
+				parent: None,
+				children: BTreeSet::new(),
+				details,
+			}
+		}
+
+		pub fn new_node(
+			hierarchy_root_id: DelegationNodeIdOf<T>,
+			parent: DelegationNodeIdOf<T>,
+			details: DelegationDetails<T>,
+		) -> Self {
+			let mut new_node = Self::new_root_node(hierarchy_root_id, details);
+			new_node.parent = Some(parent);
+
+			new_node
+		}
+
+		pub fn add_child(&mut self, child_id: DelegationNodeIdOf<T>) {
+			self.children.insert(child_id);
 		}
 	}
 
-	/// Creates a new delegation node that is a direct descendent of the
-	/// given node.
-	///
-	/// * root_id: the root node ID this node will be a child of
-	/// * parent - the parent node ID this node will be a child of
-	/// * owner: the identifier of the owner of the new delegation, i.e., the
-	///   new delegate
-	/// * permissions: the permission flags for the operations the delegate is
-	///   allowed to perform
-	pub fn new_node_child(
-		root_id: DelegationNodeIdOf<T>,
-		parent: DelegationNodeIdOf<T>,
-		owner: DelegatorIdOf<T>,
-		permissions: Permissions,
-	) -> Self {
-		DelegationNode {
-			root_id,
-			parent: Some(parent),
-			owner,
-			permissions,
-			revoked: false,
+	#[derive(Clone, Debug, Encode, Decode, PartialEq)]
+	pub struct DelegationDetails<T: Config> {
```

Comment:
why do we move the details to a separate struct?

Reply:
Because I think it's easier to self-contain whatever does not belong to a node as such, such as a parent, an ID, and the children.

Reply:
So you would like to have an abstract tree and then just some content. I wouldn't call it `DelegationTree` then but `Tree<Delegation>` maybe there is already a tree implementation that we could use. 🤔 

But i think it makes it not clearer and just adds one unnecessary step. 🤷🏻 But that is subjective.

Reply:
I don't think we need to have such a general implementation of a tree, but on the contrary I think that not having everything on the same level is a good idea.

Reply:
I think I couldn't convey my point so let me rephrase this.

In my opinion it's always a good idea to structure data when there is a category to structure them by. Eg. don't have `x` and `y` fields, but have a nested `position` field that has `y` and `x`. It makes sense to have position because it is a meaningful group for `x` and `y`.

Looking at the `DelegationDetails` it's not clear what this category is. What goes into details and what doesn't? `details` is not a real category. It's just a way of saying the number of fields in this struct was to high and i move it some where else.

But I think you actually split the fields by things that are relevant for the tree structure and things that are not relevant for the tree structure. So for me a meaningful category would be `tree_content` which makes only sense if you have a generic tree data structure.

For me it looks like the structure does more harm than it helps at the moment (especially `DelegationHierarchyDetails` which only has one field).

That said, this is my opinion and it's a style thing. So if i couldn't convince you go ahead and merge.

tl;dr; I fucking hate the word details! could be anything! What even are details?!

Reply:
I see your point, and I probably agree that details is not the best word. Plus, you got the point that my intention here is twofold:

- group everything that is tree-related and everything else that is not tree related. How could that be achieved in a different way that perhaps satisfies both preferences?
- for the hierarchy, it is true that for now it is just the CTYPE, but could be in future something else as well. And we would have to make it a struct anyway eventually, right? In order to avoid having tuples around. What would be your suggestion here?

Reply:
imHo i wouldn't have a nested structure here. 😅 If you prefer it nested, I don't see a better solution.